// @cloudscape-design/components/internal/generated/custom-css-properties/index.js
const customCSSPropertiesMap = {
    breadcrumbsGap: '--awsui-breadcrumbs-gap-lgosor',
    contentGapLeft: '--awsui-content-gap-left-lgosor',
    contentGapRight: '--awsui-content-gap-right-lgosor',
    contentHeight: '--awsui-content-height-lgosor',
    defaultMaxContentWidth: '--awsui-default-max-content-width-lgosor',
    defaultMinContentWidth: '--awsui-default-min-content-width-lgosor',
    drawerSize: '--awsui-drawer-size-lgosor',
    footerHeight: '--awsui-footer-height-lgosor',
    headerGap: '--awsui-header-gap-lgosor',
    headerHeight: '--awsui-header-height-lgosor',
    layoutWidth: '--awsui-layout-width-lgosor',
    mainGap: '--awsui-main-gap-lgosor',
    mainOffsetLeft: '--awsui-main-offset-left-lgosor',
    mainTemplateRows: '--awsui-main-template-rows-lgosor',
    maxContentWidth: '--awsui-max-content-width-lgosor',
    minContentWidth: '--awsui-min-content-width-lgosor',
    mobileBarHeight: '--awsui-mobile-bar-height-lgosor',
    notificationsGap: '--awsui-notifications-gap-lgosor',
    notificationsHeight: '--awsui-notifications-height-lgosor',
    offsetTop: '--awsui-offset-top-lgosor',
    offsetTopWithNotifications: '--awsui-offset-top-with-notifications-lgosor',
    overlapHeight: '--awsui-overlap-height-lgosor',
    navigationWidth: '--awsui-navigation-width-lgosor',
    splitPanelReportedHeaderSize: '--awsui-split-panel-reported-header-size-lgosor',
    splitPanelReportedSize: '--awsui-split-panel-reported-size-lgosor',
    splitPanelHeight: '--awsui-split-panel-height-lgosor',
    splitPanelMinWidth: '--awsui-split-panel-min-width-lgosor',
    splitPanelMaxWidth: '--awsui-split-panel-max-width-lgosor',
    toolsMaxWidth: '--awsui-tools-max-width-lgosor',
    toolsWidth: '--awsui-tools-width-lgosor',
    toolsAnimationStartingOpacity: '--awsui-tools-animation-starting-opacity-lgosor',
    contentScrollMargin: '--awsui-content-scroll-margin-lgosor',
    flashbarStackDepth: '--awsui-flashbar-stack-depth-lgosor',
    flashbarStackIndex: '--awsui-flashbar-stack-index-lgosor',
    flashbarStickyBottomMargin: '--awsui-flashbar-sticky-bottom-margin-lgosor',
    stackedNotificationsBottomMargin: '--awsui-stacked-notifications-bottom-margin-lgosor',
    stackedNotificationsDefaultBottomMargin: '--awsui-stacked-notifications-default-bottom-margin-lgosor',
    dropdownDefaultMaxWidth: '--awsui-dropdown-default-max-width-lgosor',
    spinnerRotatorFrom: '--awsui-spinner-rotator-from-lgosor',
    spinnerRotatorTo: '--awsui-spinner-rotator-to-lgosor',
    spinnerLineLeftFrom: '--awsui-spinner-line-left-from-lgosor',
    spinnerLineLeftTo: '--awsui-spinner-line-left-to-lgosor',
    spinnerLineRightFrom: '--awsui-spinner-line-right-from-lgosor',
    spinnerLineRightTo: '--awsui-spinner-line-right-to-lgosor',
};
export default customCSSPropertiesMap;
