// @cloudscape-design/components/test-utils/dom/app-layout/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
const dom_1 = require("@cloudscape-design/test-utils-core/dom");
const split_panel_1 = require("../split-panel");
const styles_selectors_js_1 = require("../../../app-layout/test-classes/styles.selectors.js");
const styles_selectors_js_2 = require("../../../split-panel/styles.selectors.js");
const button_dropdown_1 = require("../button-dropdown");
class AppLayoutWrapper extends dom_1.ComponentWrapper {
    findNavigation() {
        return this.findByClassName(styles_selectors_js_1.default.navigation);
    }
    findNavigationToggle() {
        return this.findByClassName(styles_selectors_js_1.default['navigation-toggle']);
    }
    findNavigationClose() {
        return this.findByClassName(styles_selectors_js_1.default['navigation-close']);
    }
    findContentRegion() {
        return this.findByClassName(styles_selectors_js_1.default.content);
    }
    findNotifications() {
        return this.findByClassName(styles_selectors_js_1.default.notifications);
    }
    findBreadcrumbs() {
        return this.findByClassName(styles_selectors_js_1.default.breadcrumbs);
    }
    findTools() {
        return this.findByClassName(styles_selectors_js_1.default.tools);
    }
    findToolsClose() {
        return this.findByClassName(styles_selectors_js_1.default['tools-close']);
    }
    findToolsToggle() {
        return this.findByClassName(styles_selectors_js_1.default['tools-toggle']);
    }
    findSplitPanel() {
        return this.findComponent(`.${split_panel_1.default.rootSelector}`, split_panel_1.default);
    }
    findSplitPanelOpenButton() {
        return this.findByClassName(styles_selectors_js_2.default['open-button']);
    }
    findActiveDrawer() {
        return this.findByClassName(styles_selectors_js_1.default['active-drawer']);
    }
    findActiveDrawerCloseButton() {
        return this.findByClassName(styles_selectors_js_1.default['active-drawer-close-button']);
    }
    findDrawersTriggers() {
        return this.findAllByClassName(styles_selectors_js_1.default['drawers-trigger']);
    }
    findDrawerTriggerById(id) {
        return this.find(`.${styles_selectors_js_1.default['drawers-trigger']}[data-testid="awsui-app-layout-trigger-${id}"]`);
    }
    findDrawersOverflowTrigger() {
        return this.findComponent(`.${styles_selectors_js_1.default['overflow-menu']}`, button_dropdown_1.default);
    }
    findActiveDrawerResizeHandle() {
        return this.findByClassName(styles_selectors_js_1.default['drawers-slider']);
    }
}
exports.default = AppLayoutWrapper;
AppLayoutWrapper.rootSelector = styles_selectors_js_1.default.root;
