// @cloudscape-design/component-toolkit/lib/internal/base-component/metrics/log-clients.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { __assign } from "tslib";
function validateLength(value, maxLength) {
    return !value || value.length <= maxLength;
}
/**
 * Console Platform's client logging JS API client.
 */
var CLogClient = /** @class */ (function () {
    function CLogClient() {
    }
    /**
     * Sends metric but only if Console Platform client logging JS API is present in the page.
     */
    CLogClient.prototype.sendMetric = function (metricName, value, detail) {
        if (!metricName || !/^[a-zA-Z0-9_-]+$/.test(metricName)) {
            console.error("Invalid metric name: ".concat(metricName));
            return;
        }
        if (!validateLength(metricName, 1000)) {
            console.error("Metric name ".concat(metricName, " is too long"));
            return;
        }
        if (!validateLength(detail, 4000)) {
            console.error("Detail for metric ".concat(metricName, " is too long: ").concat(detail));
            return;
        }
        var AWSC = this.findAWSC(window);
        if (typeof AWSC === 'object' && typeof AWSC.Clog === 'object' && typeof AWSC.Clog.log === 'function') {
            AWSC.Clog.log(metricName, value, detail);
        }
    };
    CLogClient.prototype.findAWSC = function (currentWindow) {
        try {
            if (typeof (currentWindow === null || currentWindow === void 0 ? void 0 : currentWindow.AWSC) === 'object') {
                return currentWindow === null || currentWindow === void 0 ? void 0 : currentWindow.AWSC;
            }
            if (!currentWindow || currentWindow.parent === currentWindow) {
                // When the window has no more parents, it references itself
                return undefined;
            }
            return this.findAWSC(currentWindow.parent);
        }
        catch (ex) {
            // Most likely a cross-origin access error
            return undefined;
        }
    };
    return CLogClient;
}());
export { CLogClient };
/**
 * Console Platform's client v2 logging JS API client.
 */
var PanoramaClient = /** @class */ (function () {
    function PanoramaClient() {
    }
    /**
     * Sends metric but only if Console Platform client v2 logging JS API is present in the page.
     */
    PanoramaClient.prototype.sendMetric = function (metric) {
        if (typeof metric.eventDetail === 'object') {
            metric.eventDetail = JSON.stringify(metric.eventDetail);
        }
        if (typeof metric.eventValue === 'object') {
            metric.eventValue = JSON.stringify(metric.eventValue);
        }
        if (!validateLength(metric.eventDetail, 4000)) {
            console.error("Event detail for metric is too long: ".concat(metric.eventDetail));
            return;
        }
        if (!validateLength(metric.eventValue, 4000)) {
            console.error("Event value for metric is too long: ".concat(metric.eventValue));
            return;
        }
        if (!validateLength(metric.eventContext, 4000)) {
            console.error("Event context for metric is too long: ".concat(metric.eventContext));
            return;
        }
        if (!validateLength(metric.eventType, 50)) {
            console.error("Event type for metric is too long: ".concat(metric.eventType));
            return;
        }
        var panorama = this.findPanorama(window);
        if (typeof panorama === 'function') {
            panorama('trackCustomEvent', __assign(__assign({}, metric), { timestamp: Date.now() }));
        }
    };
    PanoramaClient.prototype.findPanorama = function (currentWindow) {
        try {
            if (typeof (currentWindow === null || currentWindow === void 0 ? void 0 : currentWindow.panorama) === 'function') {
                return currentWindow === null || currentWindow === void 0 ? void 0 : currentWindow.panorama;
            }
            if (!currentWindow || currentWindow.parent === currentWindow) {
                // When the window has no more parents, it references itself
                return undefined;
            }
            return this.findPanorama(currentWindow.parent);
        }
        catch (ex) {
            // Most likely a cross-origin access error
            return undefined;
        }
    };
    return PanoramaClient;
}());
export { PanoramaClient };
